src/signature.c

```diff
@@ -739,6 +847,64 @@ static int p11prov_sig_digest_update(P11PROV_SIG_CTX *sigctx,
     return RET_OSSL_OK;
 }
 
+static CK_RV mech_fallback_final(P11PROV_SIG_CTX *sigctx, unsigned char *sig,
+                                 size_t *siglen, size_t sigsize, size_t mdsize)
+{
+    P11PROV_SIG_CTX *subctx;
+    CK_BYTE digest[mdsize];
+    unsigned int digest_len = mdsize;
+    CK_OBJECT_HANDLE handle;
+    int err;
+    CK_RV ret;
+
+    err = EVP_DigestFinal_ex(sigctx->mechanism_fallback, digest, &digest_len);
+    if (err != RET_OSSL_OK) {
+        ret = CKR_GENERAL_ERROR;
+        P11PROV_raise(sigctx->provctx, ret, "EVP_DigestFinal_ex() failed");
+        return ret;
+    }
+    if (digest_len != mdsize) {
+        ret = CKR_GENERAL_ERROR;
+        P11PROV_raise(sigctx->provctx, ret, "Inconsistent digest size");
+        return ret;
+    }
+
+    /* done with digest session, make sure it can't be used further */
+    p11prov_return_session(sigctx->session);
+    sigctx->session = NULL;
+
+    handle = p11prov_obj_get_handle(sigctx->key);
+    if (handle == CK_INVALID_HANDLE) {
+        ret = CKR_KEY_HANDLE_INVALID;
+        P11PROV_raise(sigctx->provctx, ret, "Provided key has invalid handle");
+        return ret;
+    }
+
+    subctx = p11prov_sig_newctx(sigctx->provctx, sigctx->mechtype,
+                                sigctx->properties);
+    if (!subctx) {
+        return CKR_HOST_MEMORY;
+    }
+
+    err = p11prov_sig_op_init(subctx, sigctx->key, sigctx->operation, NULL);
+    if (err != RET_OSSL_OK) {
+        ret = CKR_GENERAL_ERROR;
+        P11PROV_raise(sigctx->provctx, ret, "Failed to setup sigver fallback");
+        goto done;
+    }
+
+    err = p11prov_sig_operate(subctx, sig, siglen, sigsize, digest, mdsize);
+    if (err != RET_OSSL_OK) {
+        ret = CKR_GENERAL_ERROR;
+        P11PROV_raise(sigctx->provctx, ret, "Failure in sigver fallback");
+        goto done;
+    }
+
+done:
+    p11prov_sig_freectx(subctx);
+    return CKR_OK;
```

Comment:
this should be `return ret` (with setting the `ret = CKR_OK` before the label) otherwise we are returning CKR_OK even if either of the last two actions fail.

Reply:
good catch, changing to `return ret;`, and having *all* error conditions `goto done;`